### ISSUE TITLE
Possible bug in expected_returns.mean_historical_return

### ISSUE THREAD

**Issue by cstich** [robertmartin8/PyPortfolioOpt#241]
**Describe the bug**
The documentation for `mean_historical_return` says that it is only correct, when using log returns, however when just `prices` and no returns are provided it defaults to calculating the normal and not the log returns: 
`returns = returns_from_prices(prices)`

**Expected behavior**
Shouldn't it be `returns = returns_from_prices(prices, log_returns=True)`?

**Operating system and python version:**
I had a look at this commit d60271092ef8bbf7846f53cba223e7b0b52b6111

**Comment by robertmartin8** [comment]
@cstich Yup, I think the docstring is wrong on this one. I will push a fix in the next couple of weeks, along with several other improvements.

Thanks for raising this!
Robert

**Comment by SeaPea1** [comment]
@cstich @robertmartin8 
Hi Guys,

I agree, the docstring is wrong.  However, I believe the problems go a bit further.  I'll attempt to explain, would appreciate you letting me know what you think.
Firstly, let's assume you want to compound, so "compounding=True", though more on this later.
Secondly, mean_historical_return is potentially quite complex so it is easy to confuse people.  I think it makes sense to consider the cases it may make sense to support:
  1) Accept prices, return mean annualised log return.
  2) Accept prices, return mean annualised simple (i.e. "normal") return.
  3) Accept log returns, return mean annualised log return.
  4) Accept simple returns, return mean annualised simple return.
I don't personally think we want to complicate matters further by accepting log return params but return a simple returns, or vice versa.
Then a "log_returns=True" param could be added to make it clear that (1) and (3) are the default behaviours, or not add this new param and only support those.

Now getting back to the compounding param, because mean_historical_return calls returns_from_prices with the default "log_returns=False" and mean_historical_return has by default "compounding=True" the the values will be correct for case (2).
I also think that mean_historical_return will return the correct value for:
  case (4) when returns_data=True, compounding=True and "prices" are simple returns. 
  case (3) when returns_data=True, compounding=False and "prices" are log returns. 
However, this is misleading/an abuse of the compounding parameter.  I can see from the tests that it is often overridden to be False, perhaps for this reason.

Anything that changes would presumably need to be consistent in the ema_historical_return and capm_return functions too, though I haven't looked into them other than observing that all 3 functions can be invoked from return_model.

Might it make sense to remove the compounding param, given it is misleading or wrong, even though this might be a breaking change?  At least that way people would be warned that their usage was potentially wrong?

**Comment by robertmartin8** [comment]
@SeaPea1 really appreciate you taking the time! My take on the different cases is as follows:

I want the user to be able to choose one of two outputs (1) annualised arithmetic mean return, and (2) annualised geometric mean return. The idea is that `compounding=True/False` chooses between these. 

What complicates matters is that I want users to be able to provide one of three possible inputs (1) price data (2) simple returns (3) log returns. 

I've reviewed the `expected_returns.py` code again. These are all possible cases (mostly for my benefit):

a. Price data provided, `compounding=True` (default). This calculates the CAGR, probably the "correct" estimate of future returns.
b. Price data provided, `compounding=False`. Computes the annualised simple mean of historical returns (not log returns). This is technically incorrect (simple vs geometric mean), but is a common shortcut.
c. Simple returns data provided, `compounding=True`. Same as case (a). 
d. Simple returns data provided, `compounding=False`. Same as case (b).
e. Log returns data provided, `compounding=True`. This is produces something meaningless.
f. Log returns data provided, `compounding=False`. The arithmetic mean of log returns results in mean annualised log return. This is not what we want unless we exponentiate it!

So to sum up, this sentence is incorrect:

```
By default, this uses the arithmetic mean (correct if log_returns are used).
```

But I think `compounding` is acceptable. 

Looking at cases (e) and (f), we see that it is always incorrect to pass log returns to a method in `expected_returns.py`. So I could write a method `prices_from_log_returns` to go along with the existing `prices_from_returns`, such that a user with only log returns can use `prices = prices_from_log_returns(log_rets)`, then pass prices into the methods.

@SeaPea1 let me know if this makes sense, or if I am misunderstanding your criticism of the `compounding` parameter.

Cheers,
Robert

**Comment by SeaPea1** [comment]
No problem.  Sorry for taking a while to reply, real life has intervened!

Just a couple of minor points:
Regarding your (f), I think that is fine though, just depends how it is subsequently used. My point was more that you get the right answer with the wrong value for "compounding" in this case.
As to prices_from_returns, I fixed that rather than add an extra function, particularly as log_returns_from_prices is deprecated and passing a log_returns param to functions is now the preferred approach.  That's why I wondered if the same log_returns param should be added to mean_historical_return.

I agree the docstring should be corrected, maybe with more detail on the docstring for the compounding param too.

Cheers.

**Comment by robertmartin8** [comment]
@SeaPea1 

Ohh, I see what you mean; you're saying that in case (f), `compounding=False` but the output is a correctly compounded (log) return. 

My view is that in the context of PyPortfolioOpt, the output of (e) and (f) are not appropriate – they should be exponentiated before being passed into a mean-variance optimiser. 

> As to prices_from_returns, I fixed that rather than add an extra function

Oops, I misread the name of the test as a new function. But your fix has the same effect.

Here is my plan for resolving this (will be fixing bugs/ merging PRs this weekend):

- For now, I'm not going to complicate `mean_historical_return` any further by including a `log_returns_data` parameter. I would say that it's a bit of a corner case and it's not worth complicating the API – users who need this can still proceed via `prices = prices_from_returns(rets, log_returns=True)`.
- I will clarify in the docstring that `log_returns` should _not_ be passed into `mean_historical_returns`, even with `returns_data = True`. Will also clarify that `compounding` is nothing more than a toggle between arithmetic and geometric mean.

Let me know if there's anything I missed. Appreciate your time as always!

**Comment by SeaPea1** [comment]
This all seems completely reasonable to me and I think it covers everything
of interest here too.

**Comment by robertmartin8** [comment]
Fixed in v1.3.0
